caikit/runtime/service_generation/protoable.py

```diff
@@ -0,0 +1,158 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains our logic about what constitutes a proto-able for RPC generation purposes
+"""
+
+# Standard
+from typing import Dict, Type, Union, get_args, get_origin
+import inspect
+import typing
+
+# First Party
+import alog
+
+# Local
+from .type_helpers import PROTO_TYPE_MAP
+from caikit.core.data_model.base import DataBase
+
+log = alog.use_channel("PROTOABLES")
+
+
+def to_protoable_signature(signature: Dict[str, Type]) -> Dict[str, Type]:
+    """Returns dictionary of protoable types only
+    If there is a Union, pick the protoable type
+
+    Args:
+        signature: Dict[str, Type]
+            module signature of parameters and types
+    """
+    protoables = {}
+    log.debug("Building protoable signature for %s", signature)
+    for arg, arg_type in signature.items():
+        protoable_type = handle_protoables_in_union(arg_type)
+        if protoable_type:
+            protoables[arg] = protoable_type
+
+    return protoables
+
+
+def handle_protoables_in_union(arg_type: Type) -> Type:
+    """Handles various protoable arg types from a Union.
+
+    If arg_type is a union, then this will return the union back if all types in it are proto-able,
+    or the first proto-able arg type if the union has non-protoable arg types.
+
+    If arg_type is not a union nor protoable at all, this returns None.
+
+    Examples:
+    Union[protoable_type, non_protoable_type] -> protoable_type
+    Union[protoable_type_1, protoable_type_2] -> Union[protoable_type_1, protoable_type_2]
+    Union[protoable_type_1, protoable_type_2, non_protoable_type] -> protoable_type_1
+    """
+    if is_protoable_type(arg_type):
+        if typing.get_origin(arg_type) == Union:
+            union_protoables = [
+                union_val
+                for union_val in typing.get_args(arg_type)
+                if is_protoable_type(union_val)
+            ]
+            # if all are protoable, return the union (which will create a oneof)
+            if len(union_protoables) == len(typing.get_args(arg_type)):
+                return arg_type
+            # if there's only 1 protoable found, return that
+            if len(union_protoables) == 1:
+                return union_protoables[0]
+            # otherwise, try to get the data model objects in the Union
+            dm_types = [
+                arg
+                for arg in union_protoables
+                if inspect.isclass(arg) and issubclass(arg, DataBase)
+            ]
+            # if there are multiple, pick the first one
+            if len(dm_types) > 0:
```

Comment:
Ah, this logic is only for dm types. Is the idea here that `Union[RawDocument, str]` should result in just `RawDocument`?